### PR TITLE
Move all messages into a new Message component.

### PR DESCRIPTION
**Pull request by ob6160**
Also taking the opportunity to try Vue mixins which will allow us to greatly reduce code duplication.

### CODE REVIEW

File: src/components/Message.vue [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<template id="message">
+  <transition name="fade" v-if="visible">
+    <div class="message" v-bind:class="{'error': error, 'good': success}" v-html="text"></div>
+  </transition>
```
Comment (gingerbreadpup) [inline src/components/Message.vue]: maybe we could just call good success too? I think we need to get rid of the goods!

File: src/components/Queue.vue [hunk 2]
```diff
@@ -70,23 +66,20 @@ export default {
...
-                  this.spinner = false;
-              }
-
-          },
-          deep: true
-      },
-
+        queue: {
+            handler: function(val) {
+                let queueState = val.sendingStatus || false;
```
Comment (gingerbreadpup) [inline src/components/Queue.vue]: I have no idea but is let an ES5 thing?

**Review by gingerbreadpup — COMMENTED** [review]
Looks great so far. :+1:

File: src/components/Queue.vue (hunk 2 above)
Comment (ob6160) [reply src/components/Queue.vue]: This was added in another PR. But seeing as we're making most things ES5 I guess we should change it to a var. Babel will convert it to a var either way.

File: src/components/Message.vue (hunk 1 above)
Comment (ob6160) [reply src/components/Message.vue]: I agree, prefer success over good :grin:

**Review by gingerbreadpup — APPROVED** [review]
Looks good to me - Katja should probably approve too

File: src/components/Queue.vue [hunk 3]
```diff
@@ -4,17 +4,7 @@
...
-                {{ message }}
-            </div>
-            <div v-if="message" class="message">
-                {{ message }}
-            </div>
-            <div v-else class="message">
-                [xXx]You have <strong>{{ vouchers.length }}</strong> voucher<span v-if="vouchers.length > 1">s</span> in your queue. Queued vouchers will be checked when you submit your queue.
-            </div>
-        </transition>
+        <message :text="message.text" :state="message.state"></message>
```
Comment (katjam) [inline src/components/Queue.vue]: We have been explicitly using v-bind in other places.

File: src/components/Queue.vue [hunk 4]
```diff
@@ -113,68 +128,77 @@ export default {
...
+            }
+
+            if (data.invalid.length === 1) {
+                invalid = "and 1 was invalid.";
+            } else {
+                invalid = "and " + data.invalid.length + " were invalid.";
+            }
+
+            var message
+                = "Thanks! Your queue has been successfully submitted. "
```
Comment (katjam) [inline src/components/Queue.vue:172]: There was a ticket about not displaying for 0 values - but that might be another card?

File: src/constants.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export default {
```
Comment (katjam) [inline src/constants.js:1]: Not sure what we gain from this abstraction.

File: src/pages/Scan.vue [hunk 6]
```diff
@@ -7,9 +7,7 @@
                 <h1>Scan a voucher code</h1>
 
                 <form id="textVoucher" v-on:submit.prevent>
-                    <transition name="fade">
-                        <div v-if="errorMessage && (!showQueueMsg)" class="message error">{{ errorMessage }}</div>
-                    </transition>
+                    <message :text="message.text" :state="message.state"></message>
```
Comment (katjam) [inline src/pages/Scan.vue]: Explicit v-bind for consistency?

File: src/pages/Scan.vue [hunk 7]
```diff
@@ -51,7 +49,7 @@
             </div>
 
             <div>
-                <queue @update="queueMessage"></queue>
+                <queue @message-update="setMessage"></queue>
```
Comment (katjam) [inline src/pages/Scan.vue]: Again - sometimes explicit v-on sometimes @

File: src/pages/Scan.vue [hunk 8]
```diff
@@ -62,13 +60,20 @@
 <script>
 
 import Store from '../store.js';
+import mixin from '../mixins/mixins';
 import Profile from '../components/Profile.vue';
 import Queue from '../components/Queue.vue';
+import Message from '../components/Message.vue';
+
```
Comment (katjam) [inline src/pages/Scan.vue]: I bet there is a good reason for this blank line

File: src/components/Queue.vue [hunk 9]
```diff
@@ -69,24 +69,29 @@ export default {
...
-                  this.spinner = false;
-              }
-
-          },
-          deep: true
-      },
-
+        queue: {
+            handler: function(val) {
+                var queueState = val.sendingStatus || false;
```
Comment (katjam) [inline src/components/Queue.vue]: Will it be undefined if not exist?

File: src/components/Queue.vue [hunk 10]
```diff
@@ -69,24 +69,29 @@ export default {
...
+                if(!queueState && val.sentData) {
+                    var message = this.genQueueSuccessMessage(val.sentData);
+                    this.emitMessage(message, constants.MESSAGE_SUCCESS);
+
```
Comment (katjam) [inline src/components/Queue.vue]: I bet you have a good reason for this blank line.

File: src/components/Queue.vue [hunk 11]
```diff
@@ -69,24 +69,29 @@ export default {
...
-                  this.showValidate();
-              } else if(!queueState) {
-                  this.spinner = false;
-              }
-
-          },
-          deep: true
-      },
-
+        queue: {
```
Comment (katjam) [inline src/components/Queue.vue:80]: Documentation might be helpful so we know what is intended at a glance.

**Review by katjam — COMMENTED** [review]
Just little niggles - probably ok to go in as is.

File: src/components/Queue.vue (hunk 9 above)
Comment (ob6160) [reply src/components/Queue.vue]: It'll still be falsey you're right.. I'll remove the false

File: src/components/Queue.vue (hunk 4 above)
Comment (ob6160) [reply src/components/Queue.vue:172]: I just took the message as I found it. It's not related directly to the message component so it should be resolved in another card most likely.

File: src/constants.js (hunk 5 above)
Comment (ob6160) [reply src/constants.js:1]: It makes it easier to remember what 0,1,2 mean. We could alternatively pass in "error", "success", "info" when we set the message components' type - but this felt cleaner to me. Plus as I mentioned earlier we could possibly place other things such as copy into this new constants store.

File: src/components/Queue.vue (hunk 11 above)
Comment (ob6160) [reply src/components/Queue.vue:80]: Added on line 80

**Review by katjam — APPROVED** [review]
Once Travis finishes. :hibiscus:

**Pull request merged**